identity-aware-proxy/IAPClient/IAPClient.cs

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2018 Google Inc
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// [START generate_iap_request]
+
+using System;
+using System.Collections.Generic;
+using System.IdentityModel.Tokens.Jwt;
+using System.IO;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using Google.Apis.Auth.OAuth2;
+using Google.Apis.Json;
+using Microsoft.IdentityModel.Tokens;
+using Newtonsoft.Json;
+
+namespace GoogleCloudSamples
+{
+    class IAPClient
+    {
+        public static string InvokeRequest(string iapClientId,
+            string credentialsPath, string uri)
+        {
+            // Read credentials from the credentials .json file.
+            Credentials credentials;
+            using (var fs = new FileStream(credentialsPath,
+                FileMode.Open, FileAccess.Read))
+            {
+                credentials = NewtonsoftJsonSerializer.Instance
+                    .Deserialize<Credentials>(fs);
+            }
+
+            // Generate a JWT signed with the service account's private key 
+            // containing a special "target_audience" claim.
+            var jwtBasedAccessToken =
+                CreateAccessToken(credentials.PrivateKey, iapClientId,
+                    credentials.ClientEmail);
+
+            var body = new Dictionary<string, string>
+            {
+                { "assertion", jwtBasedAccessToken },
+                { "grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer"}
+            };
+
+            // Request an OIDC token for the Cloud IAP-secured client ID.
+            var httpClient = new HttpClient();
+            var httpContent = new FormUrlEncodedContent(body);
+            var result = httpClient.PostAsync(GoogleAuthConsts.OidcTokenUrl,
+                httpContent).Result;
+            var responseContent = result.Content.ReadAsStringAsync().Result;
+            int statusCode = (int)result.StatusCode;
+            if (statusCode < 200 || statusCode >= 300)
+            {
+                throw new HttpRequestException(string.Format("{0} {1}\n{2}",
```

Comment:
If using string.Format, always specify formatting culture.
See here for more details:
https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings#specifying_string_comparisons_explicitly

Reply:
Why?  I want the data to be rendered according to user's culture settings.

Reply:
I have to disagree with you here. You are sending a request to Google endpoint that may produce an error. In that case you throw exception with the details of that error. You want the details to be as they originally were intended to be sent by Google and not affected by how current culture settings are set by the user. In other words, you want all developers hitting this error to get the same message and not different messages because they have different culture settings on their machine. 
Also in general, String.Format without specifying a culture settings is Microsoft's old mistake. They have a few APIs that were exposed on .NET 1.0 and exist until today that shouldn't exist. Another similar example of bad overload they still allow is Assert.AreEqual(expected, actual). You can call that in tests without specifying a message that describes the context. People use it and then when tests fail they don't have enough information in the failure.

Reply:
I updated the code to 
```cs
                throw new HttpRequestException(string.Format(
                    CultureInfo.CurrentCulture, "{0} {1}\n{2}",
                    (int)result.StatusCode, result.ReasonPhrase,
                    responseContent));
```

If the user is in Japan, why would we not want to potentially show them 三百一?